### PR TITLE
Fix max_pool2d with ceil_mode bug

### LINKED ISSUE #45357: Bug in max_pool2d with ceil_mode under certain conditions
## 🐛 Bug

In `torch.nn.functional.max_pool2d` with `ceil_mode=True`, there are cases where maximum can be taken for a fully out-of-bounds kernel. See repro.

## To Reproduce
```
>>> x = torch.randn(1, 1, 6, 7)
>>> y = torch.nn.functional.max_pool2d(x, 1, stride=(2, 2), padding=0, ceil_mode=True)
>>> y.size()
torch.Size([1, 1, 4, 4])
>>> y
tensor([[[[ 1.5680,  1.0815, -0.5165, -1.0466],
          [ 1.2615,  1.1358, -0.0340, -0.4750],
          [ 0.7457, -1.0273,  1.0388,  0.7754],
          [   -inf,    -inf,    -inf,    -inf]]]])
```
## Expected behavior

I would expect for these parameters (`kernel_size=1, stride=(2, 2), ceil_mode=True`), we would get an output of size (3, 4) since the 4th kernel on dimension 2 would start at index 6 which is out-of-bounds.

cc @ezyang @gchanan @zou3519 @albanD @mruberry

### PR DESCRIPTION
**Pull request by heitorschueroff**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46558 Fix max_pool2d with ceil_mode bug**

This PR fixes a bug with how pooling output shape was computed.

## BC Breaking Notes
Previously, a bug in the pooling code allowed a sliding window to be entirely off bounds. Now, sliding windows must start inside the input or left padding (not right padding, see https://github.com/pytorch/pytorch/issues/46929) and may only go off-bounds if ceil_mode=True.

## Note
Updating the code to allow the sliding window to start within the right padded region would break many internal assumptions. We may address this in a future PR.

fixes #45357

TODO

- [x] Ensure existing tests are checking for the correct output size

Differential Revision: [D24633372](https://our.internmc.facebook.com/intern/diff/D24633372)

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 1]
```diff
@@ -1563,6 +1563,17 @@ void _qavg_pool_nhwc_kernel(
           float multiplier = input_scale / output_scale / divide_factor;
           int input_zero_point_m_size = -input_zero_point * size;
 
+          if (size == 0) {
```
Comment (mruberry) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:1566]: Looks like this is causing an ASAN failure and another quantization test failure. What's going on here?

File: test/test_nn.py [hunk 2]
```diff
@@ -10511,6 +10511,15 @@ def test(nonlinearity, *args, **kwargs):
         test('threshold', 3, 2)
         test('threshold', 3, 2, inplace=True)
 
+    def test_pooling_shape(self, device):
```
Comment (mruberry) [inline test/test_nn.py:10514]: Add a comment explaining what this test is for

File: test/test_nn.py [hunk 3]
```diff
@@ -10511,6 +10511,15 @@ def test(nonlinearity, *args, **kwargs):
...
+        def check(expected_out_shape, sizes, *args, **kwargs):
+            t = torch.randn(sizes, device=device)
+            self.assertEqual(torch.nn.functional.max_pool3d(t, *args, **kwargs).shape, expected_out_shape)
+
+        check((1, 1, 3, 4), (1, 1, 6, 7), kernel_size=1, stride=2, padding=0, ceil_mode=True)
```
Comment (mruberry) [inline test/test_nn.py:10519]: Add a comment explaining which test case(s) are entirely outside the bounds and what that implies for the shape, vs. those that start with a valid input, vs. those that start in a padded region.

Should one of these test cases have left padding and right padding?

File: test/test_nn.py [hunk 4]
```diff
@@ -10511,6 +10511,15 @@ def test(nonlinearity, *args, **kwargs):
         test('threshold', 3, 2)
         test('threshold', 3, 2, inplace=True)
 
+    def test_pooling_shape(self, device):
+        def check(expected_out_shape, sizes, *args, **kwargs):
+            t = torch.randn(sizes, device=device)
+            self.assertEqual(torch.nn.functional.max_pool3d(t, *args, **kwargs).shape, expected_out_shape)
```
Comment (mruberry) [inline test/test_nn.py:10517]: The issue was for max_pool2d but you're only testing max_pool3d.

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp (hunk 1 above)
Comment (heitorschueroff) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:1566]: The problem was with the AVX2 code when the kernel window is outside the non-padded region. In this case size = 0 so I added a fast-path which also avoids unnecessary loop control logic.

File: test/test_nn.py (hunk 3 above)
Comment (heitorschueroff) [reply test/test_nn.py:10519]: I'll add more test cases

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp (hunk 1 above)
Comment (mruberry) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:1566]: Is this necessary? There's no tests in the PR for this quantized code path.

File: test/test_nn.py (hunk 4 above)
Comment (heitorschueroff) [reply test/test_nn.py:10517]: max_pool2d for quantized tensors calls max_pool3d. Ideally this should test all pooling code, but since they all call the same function to compute output shape, 3D gives me more dimensions to test at once.

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp (hunk 1 above)
Comment (heitorschueroff) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: removed

File: torch/nn/modules/pooling.py [hunk 5]
```diff
@@ -45,6 +45,10 @@ class MaxPool1d(_MaxPoolNd):
     for :attr:`padding` number of points. :attr:`dilation` is the stride between the elements within the
     sliding window. This `link`_ has a nice visualization of the pooling parameters.
 
+    Note:
+        When ceil_mode=True, sliding windows are allowed to go off-bounds as long as it starts within the
```
Comment (mruberry) [inline torch/nn/modules/pooling.py:49]: "as long as it..." -> "if they start within the left padding or the input. Windows that would start in the right padded region are ignored."

File: test/test_nn.py [hunk 6]
```diff
@@ -10515,6 +10515,22 @@ def test(nonlinearity, *args, **kwargs):
...
+        # Checks output shape against expected for 1D, 2D and 3D
+        def check(expected_out_shape, sizes, *args, **kwargs):
+            for kernel in ['max', 'avg']:
+                for i in [1, 2, 3]:
+                    if hasattr(torch.nn.functional, f'{kernel}_pool{i}d'):
+                        op = getattr(torch.nn.functional, f'{kernel}_pool{i}d')
+                        t = torch.randn(sizes[:i + 2], device=device)
+                        self.assertEqual(op(t, *args, **kwargs).shape, expected_out_shape[:i + 2])
+
+        check((1, 1, 3, 3, 4), (1, 1, 5, 6, 7), kernel_size=1, stride=2, padding=0, ceil_mode=True)
```
Comment (mruberry) [inline test/test_nn.py:10530]: Would you add the exact test case from the issue, too? The test can cite the issue and test it separately at the end.

Comment (mruberry) [inline test/test_nn.py:10530]: Would you add the exact test case from the issue, too? The test can cite the issue and test it separately at the end.

**Review by mruberry — APPROVED** [review]
Just two small fixes (doc strings + exact test case from issue). Cool fix!
